spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/source/TestSparkReaderDeletes.java

```diff
@@ -139,18 +151,55 @@ protected void dropTable(String name) {
   }
 
   @Override
-  public StructLikeSet rowSet(String name, Table table, String... columns) {
-    Dataset<Row> df = spark.read()
-        .format("iceberg")
-        .load(TableIdentifier.of("default", name).toString())
-        .selectExpr(columns);
+  protected boolean countDeletes() {
+    return true;
+  }
 
-    Types.StructType projection = table.schema().select(columns).asStruct();
-    StructLikeSet set = StructLikeSet.create(projection);
-    df.collectAsList().forEach(row -> {
-      SparkStructLike rowWrapper = new SparkStructLike(projection);
-      set.add(rowWrapper.wrap(row));
-    });
+  private void setDeleteCount(long count) {
+    deleteCount = count;
+  }
+
+  @Override
+  protected long deleteCount() {
+    return deleteCount;
+  }
+
+  @Override
+  public StructLikeSet rowSet(String name, Table table, String... columns) throws IOException {
+    Schema schema = table.schema().select(columns);
+    StructType sparkSchema = SparkSchemaUtil.convert(schema);
+    Types.StructType type = schema.asStruct();
+    StructLikeSet set = StructLikeSet.create(type);
+
+    CloseableIterable<CombinedScanTask> tasks = TableScanUtil.planTasks(
+        table.newScan().planFiles(),
+        TableProperties.METADATA_SPLIT_SIZE_DEFAULT,
+        TableProperties.SPLIT_LOOKBACK_DEFAULT,
+        TableProperties.SPLIT_OPEN_FILE_COST_DEFAULT);
+
+    long delCount = 0L;
```

Comment:
This feels like we are subverting the intent of this test which is to use the DF reader method to actually read the table. I think we'll need another approach for counting deletes or possibly just explicitly state how many deletes there are in tests that need to check it. At least until we have a method of reading a table with a user readable marker for deleted rows like `_isDeleted`

Reply:
I could replace 
```
    CloseableIterable<CombinedScanTask> tasks = TableScanUtil.planTasks(
        table.newScan().planFiles(),
        TableProperties.METADATA_SPLIT_SIZE_DEFAULT,
        TableProperties.SPLIT_LOOKBACK_DEFAULT,
        TableProperties.SPLIT_OPEN_FILE_COST_DEFAULT);
```
with
```
    SparkScanBuilder scanBuilder = new SparkScanBuilder(spark, table, CaseInsensitiveStringMap.empty());
    scanBuilder.pruneColumns(sparkSchema);
    SparkScan scan = (SparkScan) scanBuilder.buildMergeOnReadScan();
    List<CombinedScanTask> tasks = scan.tasks();
```
That is the path that a Spark read takes. I am also doing most of the rest of what Spark does once we have the tasks and hand them to Spark in the form of `InputPartition`s, which is in each `InputPartition`, get a `PartitionReader` (which is a subclass of either `BatchDataReader` or `RowDataReader`) and iterate through the reader. I just don't assemble the `InternalRow`s into a `DataFrame`.

Reply:
That is the path it takes at this moment, but that's the issue here, we want this test check the output of
```    
Dataset<Row> df = spark.read()
        .format("iceberg")
        .load(TableIdentifier.of("default", name).toString())
        .selectExpr(columns);
```

Having this copy the implementation details decouples the test from what actually may be happening we change the code in the future

Reply:
Here is another option:
I'll leave `TestSparkReaderDeletes` alone, and instead create a new class that also extends `DeleteReadTests` (or even `TestSparkReaderDeletes`) and override `rowSet` there. It is perhaps also not necessary to run all the tests in `DeleteReadTests` or `TestSparkReaderDeletes` in this new class. I had also increased the parameter space of `TestSparkReaderDeletes` in order to cover all 3 file formats, but I can dial it back and distribute the increased coverage to the new class. Basically, there will be no loss of test coverage.